### DISCUSSION TITLE
Question on updating bicep resource

### DISCUSSION THREAD

**Discussion by moerketh** [renovatebot/renovate#28354] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Azure DevOps

### Please tell us more about your question or problem

I'm wondering why Renovate does not generate an update PR for resource `Microsoft.OperationalInsights/workspaces@2021-06-01` while newer version `Microsoft.OperationalInsights/workspaces@2023-09-01` should be available [see here](https://github.com/Azure/bicep-types-az/blob/main/generated/operationalinsights/microsoft.operationalinsights/2023-09-01/types.json) and present in the [index.json](https://raw.githubusercontent.com/Azure/bicep-types-az/main/generated/index.json). It seems to be working fine for other resources (for example a storageaccount).

The difference seems to be that in that index.json the storageaccount version "2023-01-01" is listed under both "resource" and "resourceFunctions" while for Microsoft.OperationalInsights/workspaces version "2023-09-01" is listed under "resources" but only version "2015-03-20" is listed under "resourceFunctions".

I'm going to make a guess here. Since I'm not seeing a [test](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/azure-bicep-resource/index.spec.ts) for this particular case (where resources and resourceFunctions have different API versions listed). Could it be that this case is not correctly handled?

I've created a minimal example with full debug log in the readme [here](https://github.com/moerketh/renovate-min-example). Notice that only a PR for the storageaccount is created not for the Log Analytics workspace.

There relavant part of the log seems to be here, where no update is found:
```
"bicep": [
      {
        "deps": [
          {
            "datasource": "azure-bicep-resource",
            "versioning": "azure-rest-api",
            "depName": "Microsoft.OperationalInsights/workspaces",
            "currentValue": "2021-06-01",
            "autoReplaceStringTemplate": "'{{depName}}@{{newValue}}'",
            "replaceString": "'Microsoft.OperationalInsights/workspaces@2021-06-01'",
            "updates": [],
            "packageName": "Microsoft.OperationalInsights/workspaces",
            "warnings": [],
            "currentVersion": "2021-06-01",
            "fixedVersion": "2021-06-01"
          },
          {
            "datasource": "azure-bicep-resource",
            "versioning": "azure-rest-api",
            "depName": "Microsoft.Storage/storageAccounts",
            "currentValue": "2022-09-01",
            "autoReplaceStringTemplate": "'{{depName}}@{{newValue}}'",
            "replaceString": "'Microsoft.Storage/storageAccounts@2022-09-01'",
            "updates": [
              {
                "bucket": "major",
                "newVersion": "2023-01-01",
                "newValue": "2023-01-01",
                "newMajor": 20230101,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/microsoft.storage-storageaccounts-20230101.x"
              }
            ],
            "packageName": "Microsoft.Storage/storageAccounts",
            "warnings": [],
            "currentVersion": "2022-09-01",
            "isSingleVersion": true,
            "fixedVersion": "2022-09-01"
          }
        ],
        "packageFile": "example.bicep"
      }
```

Am I missing something else?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
Does the new version exist in https://raw.githubusercontent.com/Azure/bicep-types-az/main/generated/index.json ?

**Reply by moerketh** [reply]
Under "resources" yes:
![image](https://github.com/renovatebot/renovate/assets/1309462/a96b2675-573e-4356-9c98-11bf3745aa37)

Under "resourceFunctions" no:
![image](https://github.com/renovatebot/renovate/assets/1309462/2a539aba-4cef-4572-a981-b4cd15759c36)
I now also notice a casing difference, not sure if relevant.

Also note that the Bicep editor in VS Code suggests the resources API's (not the resourceFunctions):
![image](https://github.com/renovatebot/renovate/assets/1309462/773c4d56-efe6-42f6-962d-a05790b04aa6)

**Reply by rarkins** [reply]
Can you experiment with the casing difference to see if that is the cause?

**Reply by moerketh** [reply]
Casing makes no difference in my minimal example testcase. I'm pretty sure this is a logic bug in Renovate and I can prove it with an added datasource test. Please forgive this crude way of sharing the code in a screenshot:
![image](https://github.com/renovatebot/renovate/assets/1309462/13fd6831-56d4-49c6-a5ca-1b6eabad6190)

Please look at [the bicep manager](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/bicep/extract.spec.ts) and notice only resources are supported, not resourcefunctions. That makes sense because you would call a function like listKeys on a resource anyway.

I'm not fluent in typescript but I'm thinking [this code](https://github.com/renovatebot/renovate/blob/771b91f69fb04ce771fd9de01bc40aa03c98b685/lib/modules/datasource/azure-bicep-resource/schema.ts#L18-L21)  overwrites the `releaseMap` result of the previous loop on resources.

I would suggest deleting that loop, the result would be more correct and pass my added test:
![image](https://github.com/renovatebot/renovate/assets/1309462/9f36a252-653d-4605-bf62-9931c4fe9e0c)

It would break the functions test, but since resourcefunction versioning is not supported by the manager and I see no case for it to be useful. Deleting the test [should return versions when package is a function](https://github.com/renovatebot/renovate/blob/771b91f69fb04ce771fd9de01bc40aa03c98b685/lib/modules/datasource/azure-bicep-resource/index.spec.ts#L31-L77) makes sense to me. Thoughts?

**Reply by rarkins** [reply]
I'm happy to accept any PR improving/fixing this. Bicep was a community contribution and I'm not sure if any maintainers are familiar with it

**Reply by moerketh** [reply]
Good stuff, I'll make a PR and try to get some eyes on it.

**Comment by moerketh** [comment]
https://github.com/renovatebot/renovate/discussions/27556 is related to this. @NicolasBissig @splitt3r do you see anything missed in [this](https://github.com/renovatebot/renovate/pull/28379) proposed fix?

**Reply by splitt3r** [reply]
I couldn't test it but it sounds good. Great work 👍🏻
